### DISCUSSION TITLE
为什么 HugePages 可以提升数据库性能 - 面向信仰编程 · /whys-the-design-linux-hugepages

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#631] [Announcements]
https://draveness.me/whys-the-design-linux-hugepages/

**Comment by ppd0705** [comment]
上次学redis专栏有了解到：fork是用了copy-on-write机制，当数据有修改，会copy一份对应的内存页，当改动比较小时比如几个字节，内存页大反而会带来更大开销

**Comment by GongchuangSu** [comment]
文章排版看着真舒服👍

**Comment by Bowser1704** [comment]
> @ppd0705
> 上次学redis专栏有了解到：fork是用了copy-on-write机制，当数据有修改，会copy一份对应的内存页，当改动比较小时比如几个字节，内存页大反而会带来更大开销

所以说 huge page 用于一些特定的场合。

> 作为 Linux 从 2.6.32 引入的新特性，HugePages 能够提升数据库、Hadoop 全家桶等占用大量内存的服务的性能，该特性对于常见的 Web 服务以及后端服务没有太多的帮助，反而可能会影响服务的性能

**Comment by meton-robean** [comment]
mmap可以申请大页面，那么这些大页面是和系统默认的其他4KB页面共存的吗？也就是说LInux可以同时使用4级查找和五级查找？

**Comment by yeqown** [comment]
> @meton-robean
> mmap可以申请大页面，那么这些大页面是和系统默认的其他4KB页面共存的吗？也就是说LInux可以同时使用4级查找和五级查找？

是共存的

**Comment by trigged** [comment]
1. `它们也不会被 Swap 到磁盘上，...` 这个说法是存疑， 参照 https://access.redhat.com/solutions/46111 
2.  在 OS 层面，要保证连续的内存同样会有压力，感觉这是一种 trade off 不是纯粹的优化  参照 Mongo https://docs.mongodb.com/manual/tutorial/transparent-huge-pages/

**Comment by draveness** [comment]
> 1. `它们也不会被 Swap 到磁盘上，...` 这个说法是存疑， 参照 https://access.redhat.com/solutions/46111
> 2. 在 OS 层面，要保证连续的内存同样会有压力，感觉这是一种 trade off 不是纯粹的优化  参照 Mongo https://docs.mongodb.com/manual/tutorial/transparent-huge-pages/

HugePages 和 THP 不是一种东西，可以看看 https://docs.oracle.com/en/database/oracle/oracle-database/12.2/ladbi/restrictions-for-hugepages-and-transparent-hugepages-configurations.html#GUID-D8178896-D00F-4F02-82A7-A44F89D8F103

**Comment by botbotty** [comment]
> 虽然部分处理器会使用 8KB、16KB 后者 64KB 作为默认的页面大小

“后者”应该是“或者”吧？

---

*2021-05-26 UPDATES: 已修复*

**Comment by hehaorui** [comment]
图3似乎有一些小问题，PGD，P4D，PUD，PMD 中表项存放的地址为下一级页表所在页面的起始地址，PTE 中存放的地址为对应物理页面的起始物理地址，所以图3中的箭头应该指向页面的起始处。至于具体的 PGD，P4D，PUD，PMD，和PTE 表项的地址需要结合虚拟地址中特定的字段计算出来。
